8263512: [macos_aarch64] issues with calling va_args functions from invoke_native

nick-arm opened the pull request:
macOS on Apple silicon uses slightly different ABI conventions to the
standard AArch64 ABI.  The differences are outlined in [1].  In
particular in the standard (AAPCS) ABI, variadic arguments may be passed
in either registers or on the stack following the normal calling
convention.  To handle this, va_list is a struct containing separate
pointers for arguments located in integer registers, floating point
registers, and on the stack.  Apple's ABI simplifies this by passing all
variadic arguments on the stack and the va_list type becomes a simple
char* pointer.

This patch adds a new MacOsAArch64 CABI type and MacOsAArch64Linker to
represent the new ABI variant on macOS.  StackVaList is based on
WinVaList lightly modified to handle the different TypeClasses on
AArch64.  The original AArch64Linker is renamed to AapcsLinker and is
currently used for all non-Mac platforms.  I think we also need to add a
WinAArch64 CABI but I haven't yet been able to test on a Windows system
so will do that later.

The macOS ABI also uses a different method of spilling arguments to the
stack (the standard ABI pads each argument to a multiple of 8 byte stack
slots, but the Mac ABI packs arguments according to their natural
alignment).  None of the existing tests exercise this so I'll open a new
JBS issue and work on that separately.

Tested jdk_foreign on macOS AArch64, Linux AArch64, and Linux X86_64.

[1] https://developer.apple.com/documentation/xcode/writing_arm64_code_for_apple_platforms

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263512](https://bugs.openjdk.java.net/browse/JDK-8263512): [macos_aarch64] issues with calling va_args functions from invoke_native


### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3617/head:pull/3617` \
`$ git checkout pull/3617`

Update a local copy of the PR: \
`$ git checkout pull/3617` \
`$ git pull https://git.openjdk.java.net/jdk pull/3617/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3617`

View PR using the GUI difftool: \
`$ git pr show -t 3617`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3617.diff">https://git.openjdk.java.net/jdk/pull/3617.diff</a>

</details>

nick-arm commented:
@mcimadamore @JornVernee could you take a look? Thanks!

JornVernee commented:
Hi Nick. Sorry for the late reply, I've been out sick. I'll hopefully be taking a thorough look at this soon (still catching up on things).

I'm pretty impressed that such a large amount of code can just be shared between the two platforms :)

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:
```diff
@@ -32,7 +32,8 @@
     SysV,
     Win64,
-    AArch64;
+    AArch64,
```
This should be Linux AArch64. Btw, should we use Arm64 instead, which seems to be the more common name to refer to the platform?

nick-arm replied:
I thought about using MacOsArm64 as that would be closer to the terminology Apple uses, but then we use AArch64 pretty consistently everywhere else in OpenJDK/Hotspot to refer to this architecture, so it seems strange to deviate here.

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/aapcs/AapcsLinker.java:
```diff
@@ -46,8 +47,8 @@
  */
-public class AArch64Linker implements CLinker {
-    private static AArch64Linker instance;
+public class AapcsLinker implements CLinker {
```
I see why it's called like that - but this is really either AArch64Linker or Arm64Linker. E.g. the platform standard one.

nick-arm replied:
How about StandardAArch64Linker? Otherwise MacOsAArch64Linker sounds like it should be a specialisation of AArch64Linker.

nick-arm replied:
Or LinuxAArch64Linker if we're going to rename the CABI enum AArch64 -> LinuxAArch64? But this would also be used on other OSs that use the Arm ABI spec as-is (e.g. *BSD).

mcimadamore commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/aapcs/AapcsVaList.java:
```diff
@@ -46,7 +47,11 @@
+ * Standard va_list implementation as defined by AAPCS document and used on
+ * Linux. Variadic parameters may be passed in registers or on the stack.
+ */
+public class AapcsVaList implements VaList {
```
Same here - not sure I like Aapcs much

mcimadamore reviewed (commented):
I'll leave fine details to @JornVernee  - high level bits seem correct. I think we should be more transparent in e.g. enum constant on the fact that AArch64 really means linux - and perhaps use Arm64 instead of AArch64.

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/macos/StackVaList.java:
```diff
@@ -0,0 +1,282 @@
+                case STRUCT_REGISTER, STRUCT_HFA -> {
+                    MemorySegment struct = allocator.allocate(layout);
+                    struct.copyFrom(segment.asSlice(0L, layout.byteSize()));
+                    segment = segment.asSlice(VA_SLOT_SIZE_BYTES);
```
Since arguments are packed according to alignment, I guess the offset could be larger or smaller than 8 bytes as well?

JornVernee replied:
This is using `alignUp(arg.layout.byteSize(), VA_SLOT_SIZE_BYTES)` in the writing code, so I think it should be the same here?

nick-arm replied:
Yes this is a mistake (`skip()` has the same problem). I'll add an extra case to VaListTest to catch it.

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/macos/StackVaList.java:
```diff
@@ -0,0 +1,282 @@
+        } else {
+            VarHandle reader = SharedUtils.vhPrimitiveOrAddress(carrier, layout);
+            res = reader.get(segment);
+            segment = segment.asSlice(VA_SLOT_SIZE_BYTES);
```
Same here I guess.

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/macos/StackVaList.java:
```diff
@@ -0,0 +1,282 @@
+ * variadic parameters are passed on the stack and the type of va_list decays to
+ * char* instead of the structure defined in the AAPCS.
+ */
+class StackVaList implements VaList {
```
VaList impls were one of the things that were affected a lot by the switch to ResourceScope, so this class will also definitely conflict with the JEP.

JornVernee commented on src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/macos/StackVaList.java:
```diff
@@ -0,0 +1,282 @@
+                return EMPTY;
+            }
+
+            // Each argument may occupy up to four slots
+            MemorySegment segment = allocator.allocate(VA_SLOT_SIZE_BYTES * args.size() * 4);
```
This seems to confirm that always offsetting by VA_SLOT_SIZE_BYTES in the read and write code can be problematic, if arguments can occupy up to 4 slots, always offsetting the read/write cursor by one slot seems incorrect.

JornVernee reviewed (commented):
I've reviewed, but I don't think this should be integrated right now since it will likely conflict with the JEP implementation, and probably a chunk of StackVaList will have to be re-written to take the ResourceScope changes into account.

As a point of process; I think this should have been a PR against the panama-foreign repo instead, which would have avoided this conflict.

Probably the best course of action right now is to wait until the JEP is integrated, and then update this PR to take those changes (e.g. ResourceScope, virtual calls) into account.

nick-arm commented:
@JornVernee thanks for the review. I'll park this until the JEP is integrated and then fix it up afterwards.

JornVernee commented:
The JEP has been integrated, so we can pick this back up (and handle it as a bug for 17 even after the fork).

Thank you for your patience.

nick-arm commented:
I've rebased this on the latest master after the JEP integration and fixed the issues above. Should be good to re-review. I've done the suggested renaming to LinuxAArch64Linker, MacOsAArch64Linker, etc.

JornVernee reviewed (approved):
Looks very good, thanks!

nick-arm commented:
Thanks @JornVernee! I noticed VaListTest has started failing on Windows with this error:
```
test VaListTest.testUpcall(java.lang.invoke.BoundMethodHandle$Species_LLLLLLL@198ebce4, MethodHandle(VaList)void): success
test VaListTest.testUpcall(java.lang.invoke.BoundMethodHandle$Species_LLLLLLL@7a97cd30, MethodHandle(VaList)void): success
Uncaught exception:
java.lang.IllegalArgumentException 
{0x00000000d6506c20} - klass: 'java/lang/IllegalArgumentException'
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (universalUpcallHandler.cpp:113), pid=13972, tid=23500
#  Error: ShouldNotReachHere()
#
```

I guess it must be related to the two new cases I added and the Windows code is now throwing an IllegalArgumentException but I can't see where from. Any ideas?

JornVernee commented:
Hey @nick-arm I'm on Windows, so I can take a look here. We recently added a patch that handles these exceptions better and actually prints the stack trace as well.

JornVernee commented on test/jdk/java/foreign/valist/VaListTest.java:
```diff
@@ -674,6 +693,17 @@ public void testCopyUnusableAfterOriginalClosed(Function<Consumer<VaList.Builder
                     assertEquals((long) VH_BigPoint_x.get(struct), 8);
                     assertEquals((long) VH_BigPoint_y.get(struct), 16);
                 })},
+                { linkVaListCB("upcallBigStructPlusScalar"), VaListConsumer.mh(vaList -> {
+                    MemorySegment struct = vaList.vargAsSegment(BigPoint_LAYOUT, ResourceScope.newImplicitScope());
+                    assertEquals((long) VH_BigPoint_x.get(struct), 8);
+                    assertEquals((long) VH_BigPoint_y.get(struct), 16);
+
+                    assertEquals((long) vaList.vargAsLong(C_LONG), 42);
+                })},
+                { linkVaListCB("upcallBigStructPlusScalar"), VaListConsumer.mh(vaList -> {
+                    vaList.skip(BigPoint_LAYOUT);
+                    assertEquals((long) vaList.vargAsLong(C_LONG), 42);
+                })},
```
Looks like a carrier size mismatch here:

```
java.lang.IllegalArgumentException: Carrier size mismatch: long != b32[abi/kind=LONG]
        at jdk.incubator.foreign/jdk.internal.foreign.Utils.checkPrimitiveCarrierCompat(Utils.java:111)
        at jdk.incubator.foreign/jdk.internal.foreign.abi.SharedUtils.checkCompatibleType(SharedUtils.java:231)
        at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.windows.WinVaList.read(WinVaList.java:114)
        at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.windows.WinVaList.read(WinVaList.java:109)
        at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.windows.WinVaList.vargAsLong(WinVaList.java:84)
        at VaListTest.lambda$upcalls$58(VaListTest.java:705)
        at jdk.incubator.foreign/jdk.internal.foreign.abi.ProgrammableInvoker.invokeNative(Native Method)
        at jdk.incubator.foreign/jdk.internal.foreign.abi.ProgrammableInvoker.invokeMoves(ProgrammableInvoker.java:290)
        at VaListTest.testUpcall(VaListTest.java:530)
```

Need to use `C_LONG_LONG` to be cross-platform compatible. (sorry for not noticing this).

```suggestion
                { linkVaListCB("upcallBigStructPlusScalar"), VaListConsumer.mh(vaList -> {
                    MemorySegment struct = vaList.vargAsSegment(BigPoint_LAYOUT, ResourceScope.newImplicitScope());
                    assertEquals((long) VH_BigPoint_x.get(struct), 8);
                    assertEquals((long) VH_BigPoint_y.get(struct), 16);

                    assertEquals((long) vaList.vargAsLong(C_LONG_LONG), 42);
                })},
                { linkVaListCB("upcallBigStructPlusScalar"), VaListConsumer.mh(vaList -> {
                    vaList.skip(BigPoint_LAYOUT);
                    assertEquals((long) vaList.vargAsLong(C_LONG_LONG), 42);
                })},
```

JornVernee replied:
Also, it looks like the cast to `(long)` on those lines is redundant (thanks IntelliJ).

nick-arm replied:
Thanks for your help! These are both fixed now.

mcimadamore reviewed (commented):
Looks good - except for the issue raised by Jorn (carrier mismatch on test)

mcimadamore commented:
@nick-arm, I've just integrated a fix which, I believe will create a minor build issue with the changes in this patch:

https://git.openjdk.java.net/jdk/pull/4316

That fix has a switch on the ABI type in the SystemLookup class (a new class introduced by that fix). I believe that switch will no longer compile with the changes in this PR as the ABI enum constants have changed - hopefully the fix should be easy - e.g. just replace the case label with references AARCH64 with a compound case label which covers both ARM/linux and ARM/MacOS.

nick-arm commented:
> 
> That fix has a switch on the ABI type in the SystemLookup class (a new class introduced by that fix). I believe that switch will no longer compile with the changes in this PR as the ABI enum constants have changed - hopefully the fix should be easy - e.g. just replace the case label with references AARCH64 with a compound case label which covers both ARM/linux and ARM/MacOS.

OK sure, the last commit fixes that.

nick-arm commented:
/integrate
